File: mqtt-example.js
#!/usr/bin/env node
/**
 * AWS IoT Core MQTT接続サンプル (Node.js)
 * 
 * このスクリプトは、create-iot-thing.shで生成された証明書を使用して
 * AWS IoT CoreにMQTT接続し、メッセージの送受信を行います。
 * 
 * 必要なライブラリ:
 *   npm install aws-iot-device-sdk
 */

const awsIot = require('aws-iot-device-sdk');
const path = require('path');

// コマンドライン引数の処理
const args = process.argv.slice(2);
if (args.length < 2 || args.includes('--help')) {
  console.log('使用方法:');
  console.log('  node mqtt-example.js <endpoint> <thing-name> [cert-dir] [topic]');
  console.log('');
  console.log('例:');
  console.log('  node mqtt-example.js xxxxx.iot.ap-northeast-1.amazonaws.com my-iot-thing');
  console.log('  node mqtt-example.js xxxxx.iot.ap-northeast-1.amazonaws.com my-iot-thing ./certificates test/topic');
  process.exit(args.includes('--help') ? 0 : 1);
}

const endpoint = args[0];
const thingName = args[1];
const certDir = args[2] || './certificates';
const topic = args[3] || 'test/topic';

// 証明書ファイルのパス
const keyPath = path.join(certDir, `${thingName}-private.pem.key`);
const certPath = path.join(certDir, `${thingName}-certificate.pem.crt`);
const caPath = path.join(certDir, 'AmazonRootCA1.pem');

console.log('======================================');
console.log('AWS IoT Core MQTT接続を開始します');
console.log('======================================');
console.log(`エンドポイント: ${endpoint}`);
console.log(`Thing名: ${thingName}`);
console.log(`トピック: ${topic}`);
console.log('');

// デバイス接続の設定
const device = awsIot.device({
  keyPath: keyPath,
  certPath: certPath,
  caPath: caPath,
  clientId: thingName,
  host: endpoint
});

// 接続イベント
device.on('connect', () => {
  console.log('✓ AWS IoT Coreに接続しました');
  console.log('');
  
  // トピックの購読
  device.subscribe(topic);
  console.log(`✓ トピック '${topic}' を購読しました`);
  console.log('');
  
  // 定期的にメッセージを送信
  let messageCount = 0;
  setInterval(() => {
    messageCount++;
    const payload = {
      message: `Hello from ${thingName}`,
      count: messageCount,
      timestamp: new Date().toISOString()
    };
    
    console.log(`メッセージを送信: ${JSON.stringify(payload)}`);
    device.publish(topic, JSON.stringify(payload));
  }, 5000);
});

// メッセージ受信イベント
device.on('message', (receivedTopic, payload) => {
  console.log('');
  console.log('メッセージを受信:');
  console.log(`  トピック: ${receivedTopic}`);
  console.log(`  メッセージ: ${payload.toString()}`);
  console.log('');
});

// エラーイベント
device.on('error', (error) => {
  console.error('エラーが発生しました:', error);
});

// 切断イベント
device.on('close', () => {
  console.log('接続が切断されました');
});

// オフラインイベント
device.on('offline', () => {
  console.log('オフラインになりました');
});

// 再接続イベント
device.on('reconnect', () => {
  console.log('再接続中...');
});

// Ctrl+Cでの終了処理
process.on('SIGINT', () => {
  console.log('');
  console.log('終了します...');
  device.end();
  process.exit(0);
});

console.log('接続中...');

File: mqtt-example.py
#!/usr/bin/env python3
"""
AWS IoT Core MQTT接続サンプル

このスクリプトは、create-iot-thing.shで生成された証明書を使用して
AWS IoT CoreにMQTT接続し、メッセージの送受信を行います。

必要なライブラリ:
    pip install AWSIoTPythonSDK

注意: このサンプルはAWSIoTPythonSDK (v1)を使用しています。
新しいプロジェクトの場合は、AWS IoT Device SDK v2 (awsiotsdk)の使用を推奨します。
詳細: https://github.com/aws/aws-iot-device-sdk-python-v2
"""

import json
import time
import argparse
from AWSIoTPythonSDK.MQTTLib import AWSIoTMQTTClient


def on_message_callback(client, userdata, message):
    """メッセージ受信時のコールバック関数"""
    print(f"受信: トピック={message.topic}")
    print(f"      メッセージ={message.payload.decode('utf-8')}")
    print()


def main():
    parser = argparse.ArgumentParser(
        description='AWS IoT Core MQTT接続サンプル'
    )
    parser.add_argument(
        '--endpoint',
        required=True,
        help='AWS IoTエンドポイント (例: xxxxx.iot.ap-northeast-1.amazonaws.com)'
    )
    parser.add_argument(
        '--thing-name',
        default='my-iot-thing',
        help='Thing名 (デフォルト: my-iot-thing)'
    )
    parser.add_argument(
        '--cert-dir',
        default='./certificates',
        help='証明書ディレクトリ (デフォルト: ./certificates)'
    )
    parser.add_argument(
        '--topic',
        default='test/topic',
        help='MQTTトピック (デフォルト: test/topic)'
    )
    
    args = parser.parse_args()
    
    # 証明書ファイルのパス
    root_ca = f"{args.cert_dir}/AmazonRootCA1.pem"
    certificate = f"{args.cert_dir}/{args.thing_name}-certificate.pem.crt"
    private_key = f"{args.cert_dir}/{args.thing_name}-private.pem.key"
    
    # MQTTクライアントの初期化
    print("AWS IoT Core MQTT接続を開始します...")
    print(f"  エンドポイント: {args.endpoint}")
    print(f"  Thing名: {args.thing_name}")
    print(f"  トピック: {args.topic}")
    print()
    
    mqtt_client = AWSIoTMQTTClient(args.thing_name)
    mqtt_client.configureEndpoint(args.endpoint, 8883)
    mqtt_client.configureCredentials(root_ca, private_key, certificate)
    
    # 接続設定
    mqtt_client.configureAutoReconnectBackoffTime(1, 32, 20)
    mqtt_client.configureOfflinePublishQueueing(-1)  # 無制限
    mqtt_client.configureDrainingFrequency(2)  # 2Hz
    mqtt_client.configureConnectDisconnectTimeout(10)  # 10秒
    mqtt_client.configureMQTTOperationTimeout(5)  # 5秒
    
    # 接続
    print("接続中...")
    mqtt_client.connect()
    print("✓ 接続成功！")
    print()
    
    # トピックの購読
    print(f"トピック '{args.topic}' を購読します...")
    mqtt_client.subscribe(args.topic, 1, on_message_callback)
    print("✓ 購読開始")
    print()
    
    # メッセージの送信
    try:
        message_count = 0
        while True:
            message_count += 1
            payload = {
                'message': f'Hello from {args.thing_name}',
                'count': message_count,
                'timestamp': time.time()
            }
            
            print(f"メッセージを送信します: {payload}")
            mqtt_client.publish(
                args.topic,
                json.dumps(payload),
                1
            )
            print("✓ 送信完了")
            print()
            
            time.sleep(5)  # 5秒待機
            
    except KeyboardInterrupt:
        print("\n終了します...")
    finally:
        mqtt_client.disconnect()
        print("✓ 切断完了")


if __name__ == '__main__':
    main()
